### ISSUE TITLE
Ability to run unit tests against unbundled source files instead of bundled files

### ISSUE THREAD

**Issue by chuckwatson87** [aurelia/cli#328]
**I'm submitting a feature request**

* **Library Version:**
0.19.0

**Please tell us about your environment:**
* **Operating System:**
Windows 10

* **Node Version:**
4.2.4

* **NPM Version:**
3.9.5

* **Browser:**
all 

* **Language:**
all

**Current behavior:**
Currently when running `au test` or `au test --watch`, the tests are being run against the bundled code.

**Expected/desired behavior:**

* **What is the expected behavior?**
to be able run tests against transpiled source code  instead of bundled code

* **What is the motivation / use case for changing the behavior?**
Currently, running the tests against the bundles is causing us to have to manually rebuild every time we make a change to the source files that we are testing.  Also makes debugging a bit harder, since we are only seeing the transpiled, bundled code in the browser when running tests.  Being able to test against the source code without bundling would improve this workflow.

**Comment by AStoker** [comment]
+1 for this feature

**Comment by Jenselme** [comment]
👍 I'd love to see that too.

**Comment by chuckwatson87** [comment]
For anyone who is interested, worked on this a little bit more yesterday and today, and was able to get tests running against unbundled source code in the skeleton project using the following config:

```javascript
"use strict";
const path = require('path');
const project = require('./aurelia_project/aurelia.json');

let testSrc = [
  { pattern: 'node_modules/**/*.js', included: false },
  { pattern: 'node_modules/**/*.html', included: false },
  { pattern: 'node_modules/**/*.css', included: false },
  { pattern: 'node_modules/**/*.map', included: false },
  { pattern: 'node_modules/**/*.woff', included: false },
  { pattern: 'node_modules/**/*.png', included: false },
  { pattern: project.unitTestRunner.source, included: false },
  { pattern: project.transpiler.source, included: false },

  'test/aurelia-karma.js'
];

//let output = project.platform.output;
let entryPoint = "scripts/require.js";

//let appSrc = project.build.bundles.map(x => path.join(output, x.name));
// let entryIndex = appSrc.indexOf(path.join(output, project.build.loader.configTarget));
// let entryBundle = appSrc.splice(entryIndex, 1)[0];
let files = [entryPoint].concat(testSrc);

module.exports = function(config) {
  config.set({
    basePath: '',
    frameworks: [project.testFramework.id],
    files: files,
    exclude: ['node_modules/**/*spec.js', 'node_modules/**/*Spec.js'],
    preprocessors: {
      [project.unitTestRunner.source]: [project.transpiler.id],
      [project.transpiler.source]: [project.transpiler.id]
    },
    'babelPreprocessor': { options: karmaOptions() },
    reporters: ['progress'],
    proxies: {
        '/src/': '/base/src/',
        '/node_modules/': '/base/node_modules/'
    },
    port: 9876,
    colors: true,
    logLevel: config.LOG_INFO,
    autoWatch: true,
    browsers: ['Chrome'],
    singleRun: false
  });
};

function karmaOptions() {
    var options = project.transpiler.options;
    options.sourceMap = 'inline';
    return options;
}
```

And then adding the following requirejs.config call to the aurelia-karma file at line 70.
```javascript
  function requireTests() {
    var TEST_REGEXP = /(spec)\.js$/i;
    var allTestFiles = ['/base/test/unit/setup.js'];

    Object.keys(window.__karma__.files).forEach(function(file) {
      if (TEST_REGEXP.test(file)) {
        allTestFiles.push(file);
      }
    });

    //call requirejs.config to add paths to necessary files:
    requirejs.config(
        {
            "paths": {
              "aurelia-pal-browser":"../node_modules\\aurelia-pal-browser\\dist\\amd\\aurelia-pal-browser",
              "aurelia-polyfills":"../node_modules\\aurelia-polyfills\\dist\\amd\\aurelia-polyfills",
              "aurelia-pal":"../node_modules\\aurelia-pal\\dist\\amd\\aurelia-pal"
            }
        });

    require(allTestFiles, window.__karma__.start);
  }
```

**Comment by AStoker** [comment]
@EisenbergEffect, what would be helpful in this case is having the `requirejs.config` in a location that can be copied and used here (instead of inside the vendor bundle). Is that something that we could implement and have as an option under the `build/loader` options? Would you see any problems with that?

**Comment by EisenbergEffect** [comment]
You mean having the CLI generate an extra file with that config into the test folder or something like that?

**Comment by AStoker** [comment]
It would generate that file into wherever the output is. And then instead of the config being integrated directly into the `vendor-bundle.js` file, it would be its own file, like `loader-config.js`.

**Comment by EisenbergEffect** [comment]
I'll need to be convinced more on that because that now adds additional script configs that are needed to run the app. I would really like to have a single script tag in the index.html.

**Comment by AStoker** [comment]
That would still be the desire. If the vendor bundle itself could load up the file, then there wouldn't need to be an additional script tag.

**Comment by AStoker** [comment]
The benefit of having that external config file is that we can have some finer control of what files are loaded and how in situations that aren't for production (such as various testing libraries). In our particular set up, we're not wanting to run tests off of bundled code, but rather the source code. Currently, that requires requireJS to be configured, but that config file isn't generated except by bundling code, and even then it's put inside the bundle itself.

**Comment by EisenbergEffect** [comment]
@AStoker I'd like to see a more detailed proposal. What you are proposing is a major breaking change to the CLI so I need to weigh that against other options and understand the consequences, especially in terms of how this might relate to future support for system.js and webpack.

**Comment by AStoker** [comment]
@EisenbergEffect, here's what I'm proposing. It's different than what I had said previously, but I think it will not have any effect on the users and how they deal with setting up their site. It will also have the benefit of separating out some concerns so that in the future, more module loaders will be easier to implement. Description as follows.  

What is needed: Test driven development
---
(in other words, run tests from source code)  
In order to have test driven development (tdd), tests must run near instantly after code changes to the source code (the part in development), providing the developer with quick feedback about his/her code change. With the current setup, test are run against bundled code, therefore any changes to the source code means _the entire application needs to be re-bundled_ (for large applications, this can be upwards of 2 minutes, which is unacceptable for tdd).
However, there is absolutely still a prominent case for running tests against the bundled code. When we publish our code to our server, we want to run tests on what the client is going to be downloading, so we want to run tests against bundled code. The issue is for the _development_ part of running tests.  

How to have TDD
---
To ensure that the time between a file being modified and the tests are run is minimal, we need to make sure that we're only getting and transpiling the files that need to be run. In other words, we need to be able to just pull in and transpile the files we need, and only when we need it. When we run tests with bundled code, we're transpiling _everything_, even things we don't need. 
In order to get this kind of speed, we need to reference the source files instead of bundled files, and rely on the babel preprocessor in karma to transpile the files we ask for. @chuckwatson87 has demonstrated this to work (with a few caveats that I'll get to next).  

The current problem
---
In order for us to have been able to pull in the source files, we require a loader config to be created so that our tests still know where to get the files. Currently, @chuckwatson87 has done this by stripping out the `requirejs.config` from the generated vendor-bundle, and then took out the `bundles` property. This resulted in a loader config that allowed karma to run tests against the source code instead of the bundled code.  
**So what we need is the ability to have a correctly generated loader config**, so that we can set up karma the way we need it to. Furthermore, if users are using Wallaby, they'd need their own config for that as well. Everything we need to correctly generate a loader config should be inside the `aurelia.json` file, so we should be able to generate a loader config in the amount of time we need to have tdd.  

Proposed changes
---
If we were to abstract process of creating the loader config _out of the `bundle.js`_ ([line 208](https://github.com/aurelia/cli/blob/master/lib/build/bundle.js#L208)), and move that process to a separate module that was made public, we would achieve two goals.  
   1. Separation of how a loader configuration is created from what the bundle is. This would allow us to more easily add loader configurations in the future (and even open the possibility of custom configuration generators).
   2. Allow the developer to call a function that would provide them with a generated loader config to use/modify for their tests.   
These changes would not effect the current development process at all (no current configurations would break), it would simply open up another avenue for running tests.  

So there you have it, a _more detailed proposal_ 😉. Quite possibly a bit too detailed, but I wanted to get as many of my thoughts out there as possible (to open up for picking them apart, or improving upon).  
I'd like to start working on a PR for this feature, and I don't think it would be terribly difficult to implement. What are your thoughts?  

### TL;DR
Let's move the function that creates the loader config out of the `bundle.js` file so that we can open it up and allow developers to use it to create a custom loader config to run tests for development.

**Comment by cResults** [comment]
:+1: +1

**Comment by EisenbergEffect** [comment]
@AStoker This sounds like a good plan. I was hoping to extract the loader specific parts anyways pretty soon so that we could support multiple loaders more easily and make that part of the system extensible. It sounds like we can kill two birds with one stone here. Please proceed with the work whenever you have time and ping me on the PR so we can start the review discussion there. Thanks!

**Comment by AStoker** [comment]
@EisenbergEffect, could you review the PR for this (https://github.com/aurelia/cli/pull/353). It works in a large scale project, but I'm still getting the karma config set up so that it can make use of the revealed functions. In the mean time though, I'd appreciate some review on the approach I too.

**Comment by EisenbergEffect** [comment]
Bookmarking for later this evening.

**Comment by Jenselme** [comment]
@AStoker Can we have a small example of how to use the work done in #353 to run the tests on the source files?

**Comment by AStoker** [comment]
@Jenselme, I'm working on getting a configuration working based off the CLI. I'll let you know when I get that working

**Comment by Jenselme** [comment]
@AStoker Looking forward to it! If you need help or a tester, let me know, I should be able to give you a hand.

**Comment by AStoker** [comment]
@Jenselme, if you're interested, I made a repo with the CLI and added modifications/notes on how to run TDD. It is dependent on getting the next version where you can externally configure the loader, but hopefully that gets deployed tomorrow. 
https://github.com/AStoker/Aurelia-TDD
